Add FormBuilder forms to Civiimport

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Add FormBuilder forms to Civiimport

Before
----------------------------------------
Without FormBuilder wrappers we can't embed the forms

After
----------------------------------------
Now we can...

Technical Details
----------------------------------------

Comments
----------------------------------------

@colemanw  - these are created if Civiimport is enabled when importing

eileenmcnaughton committed:
Add FormBuilder forms to Civiimport

Co-authored-by: colemanw <coleman@civicrm.org>

eileenmcnaughton commented:
Unrelated fail 

[Test Result](https://test.civicrm.org/job/CiviCRM-Core-PR/52542/testReport/) (2 failures / ±0)
[api_v3_OrderTest.testAddOrderForMembershipWithDates with data set #6](https://test.civicrm.org/job/CiviCRM-Core-PR/52542/testReport/junit/(root)/api_v3_OrderTest/testAddOrderForMembershipWithDates_with_data_set__6/)
[api_v3_OrderTest.testAddOrderForMembershipWithDates with data set #13](https://test.civicrm.org/job/CiviCRM-Core-PR/52542/testReport/junit/(root)/api_v3_OrderTest/testAddOrderForMembershipWithDates_with_data_set__13/)

colemanw commented on ext/civiimport/Civi/Api4/Event/Subscriber/ImportSubscriber.php:
```diff
@@ -112,4 +112,70 @@ public function onApiAuthorize(AuthorizeEvent $event): void {
+   * @noinspection PhpUnused
+   */
+  public static function on_civi_afform_get(GenericHookEvent $event): void {
+    // Early return if forms are not requested -  copied from eck code - correct?
+    if ($event->getTypes && !in_array('form', $event->getTypes, TRUE)) {
```
```suggestion
    // We're only providing afforms of type 'search'
    if ($event->getTypes && !in_array('search', $event->getTypes, TRUE)) {
```

colemanw commented on ext/civiimport/Civi/Api4/Event/Subscriber/ImportSubscriber.php:
```diff
@@ -112,4 +112,70 @@ public function onApiAuthorize(AuthorizeEvent $event): void {
+    }
+    catch (UnauthorizedException $e) {
+      // No access - return the empty array.
+    }
```
@eileenmcnaughton unless I'm missing something, all this can be simplified. You don't need the per-user caching & you don't need the permission check when fetching SearchDisplays and you don't need the try/catch.

eileenmcnaughton replied:
@colemanw yeah - you do want all that - the user will only retrieve their own jobs by the acls - so it is per user. And if they don't have permission to the api at all then we want to catch the error

colemanw replied:
Ok so is there some hidden mechanism checking permissions inside the  `SearchDisplay::get()` that only returns certain searches per-user?

eileenmcnaughton replied:
Yeah - cos it applies ACLs & the ACLS on the UserJob filter

colemanw commented on ext/civiimport/Civi/Api4/Event/Subscriber/ImportSubscriber.php:
```diff
@@ -112,4 +112,70 @@ public function onApiAuthorize(AuthorizeEvent $event): void {
+          'is_dashlet' => FALSE,
+          'is_public' => FALSE,
+          'is_token' => FALSE,
+          'permission' => 'access CiviCRM',
+          'server_route' => 'civicrm/imports/' . $importSearch['name'],
```
I think these ought to be removed since the form is meant to be embedded not on its own page.

eileenmcnaughton replied:
Done

eileenmcnaughton commented:
@colemanw I've addressed your feedback (and confirmed the docs step-by-step guide on how to embed a form is inadequate.... https://docs.civicrm.org/dev/en/latest/afform/overview/#embedding )

colemanw commented:
@eileenmcnaughton I've expanded the docs to be slightly more adequate. The code snippets are untested so will need to be adjusted as we go along.

eileenmcnaughton commented:
@colemanw - I'll see if I can make sense of it - this will be less painful to work on if we put 'My imports' in navigation - do you think under 'Reports' - it's the best I can think of

colemanw commented:
Ok @eileenmcnaughton I'm fine with you merging this whenever you're ready

eileenmcnaughton commented:
@colemanw yeah - I'm ready on this one - the other is slated as a follow up (embedding)

eileenmcnaughton commented:
@colemanw OK - having sume success with - https://github.com/civicrm/civicrm-core/pull/25087/commits/301c8deb2a1a1f088912402143c1d76f26fc856e
